tests/common/hyperparameter_tuner/test_hyperparameter_tuner.py

```diff
@@ -0,0 +1,663 @@
+                # x, y
+                (0, 0): {},
+                (0, 1): {"y": True},
+                (0, 2): {"y": False},
+                (1, 0): {"x": 0},
+                (1, 1): {"x": 0, "y": True},
+                (1, 2): {"x": 0, "y": False},
+                (2, 0): {"x": 1},
+                (2, 1): {"x": 1, "y": True},
+                (2, 2): {"x": 1, "y": False},
+                (3, 0): {"x": 2},
+                (3, 1): {"x": 2, "y": True},
+                (3, 2): {"x": 2, "y": False},
+            },
+        ),
+        # three_parameters
+        (
+            # param_settings
+            {"x": [0, 1], "y": [True, False], "z": ["a", "b"]},
+            # expected_combinations
+            {
+                # x
+                (0,): {},
+                (1,): {"x": 0},
+                (2,): {"x": 1},
+                # x, y
+                (0, 0): {},
+                (0, 1): {"y": True},
+                (0, 2): {"y": False},
+                (1, 0): {"x": 0},
+                (1, 1): {"x": 0, "y": True},
+                (1, 2): {"x": 0, "y": False},
+                (2, 0): {"x": 1},
+                (2, 1): {"x": 1, "y": True},
+                (2, 2): {"x": 1, "y": False},
+                # x, y, z
+                (0, 0, 0): {},
+                (0, 0, 1): {"z": "a"},
+                (0, 0, 2): {"z": "b"},
+                (0, 1, 0): {"y": True},
+                (0, 1, 1): {"y": True, "z": "a"},
+                (0, 1, 2): {"y": True, "z": "b"},
+                (0, 2, 0): {"y": False},
+                (0, 2, 1): {"y": False, "z": "a"},
+                (0, 2, 2): {"y": False, "z": "b"},
+                (1, 0, 0): {"x": 0},
+                (1, 0, 1): {"x": 0, "z": "a"},
+                (1, 0, 2): {"x": 0, "z": "b"},
+                (1, 1, 0): {"x": 0, "y": True},
+                (1, 1, 1): {"x": 0, "y": True, "z": "a"},
+                (1, 1, 2): {"x": 0, "y": True, "z": "b"},
+                (1, 2, 0): {"x": 0, "y": False},
+                (1, 2, 1): {"x": 0, "y": False, "z": "a"},
+                (1, 2, 2): {"x": 0, "y": False, "z": "b"},
+                (2, 0, 0): {"x": 1},
+                (2, 0, 1): {"x": 1, "z": "a"},
+                (2, 0, 2): {"x": 1, "z": "b"},
+                (2, 1, 0): {"x": 1, "y": True},
+                (2, 1, 1): {"x": 1, "y": True, "z": "a"},
+                (2, 1, 2): {"x": 1, "y": True, "z": "b"},
+                (2, 2, 0): {"x": 1, "y": False},
+                (2, 2, 1): {"x": 1, "y": False, "z": "a"},
+                (2, 2, 2): {"x": 1, "y": False, "z": "b"},
+            },
+        ),
+    ],
+    ids=[
+        "one_parameter",
+        "two_parameters",
+        "three_parameters",
+    ],
+)
+def test_create_combinations(
+    param_settings: Dict[str, List[Any]], expected_combinations: Dict[CombinationKey, Combination]
+):
+    actual_combinations = create_combinations(param_settings)
+    assert expected_combinations == actual_combinations
```

Comment:
As I can see, `actual_combinations` should be the same as `expected_combinations`.
As I can see, `expected_combinations` contains `# x` and `# x, y` combinations that are the subsamples of the `# x, y, z` combination.  The main question is: why do we need `# x` and `# x, y` combinations if we would have them in `# x, y, z`?
I mean, isn't it the problem of the `create_combinations` method implementation that checks the same combinations twice or even thrice?

Reply:
Agreed. The `create_combinations()` was refactored.